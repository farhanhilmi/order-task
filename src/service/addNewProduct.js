import productRepo from '../repository/productRepo.js';

const addNewProduct = async (data) => {
  try {
    return await productRepo.addProduct(data);
  } catch (err) {
    return err;
  }
};

export default addNewProduct;
