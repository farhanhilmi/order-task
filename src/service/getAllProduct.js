import productRepo from '../repository/productRepo.js';

const getProducts = async (call, callback) => {
  try {
    const products = await productRepo.getListProduct();
    if (!products) callback(null, { products: [] });
    callback(null, { products });
  } catch (err) {
    callback(err);
  }
};

export default getProducts;
