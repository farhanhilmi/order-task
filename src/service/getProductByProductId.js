import productRepo from '../repository/productRepo.js';

const getProductById = async (id) => {
  try {
    return await productRepo.getProductById(id);
  } catch (err) {
    return err;
  }
};

export default getProductById;
