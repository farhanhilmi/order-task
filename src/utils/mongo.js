import mongoose from 'mongoose';

export default {
  toObjectId: (id) => mongoose.Types.ObjectId(id),
  objectId: mongoose.Types.ObjectId,
  generateObjectId: () => new mongoose.Types.ObjectId(),
};
