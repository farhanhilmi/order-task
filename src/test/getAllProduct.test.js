import { jest } from '@jest/globals';
import mongo from '../utils/mongo.js';

import getAllProduct from '../service/getAllProduct.js';
import getProductByProductId from '../service/getProductByProductId.js';
import addNewProduct from '../service/addNewProduct.js';

describe('Product Service | Add New Product', () => {
  it('should store a new product with correct property', async () => {
    const newProduct = {
      name: 'apple',
      quantity: 50,
      description: 'ladkald',
      price: 10000,
    };
    const product = await addNewProduct(newProduct);
    expect(product).toMatchSnapshot();
    expect(product.name).toBeTruthy();
    expect(product.quantity).toBeTruthy();
    expect(product.description).toBeTruthy();
    expect(product.price).toBeTruthy();
  });

  it('should return an error message due to missing field', async () => {
    const newProduct = {
      name: 'apple',
      quantity: 50,
      // description: 'ladkald',
      price: 10000,
    };
    const products = await addNewProduct(newProduct);
    const throwError = () => {
      throw new TypeError(products);
    };
    expect(throwError).toThrow(`Error: description is a required field`);
  });

  it('should return an error message due to wrong input type', async () => {
    const newProduct = {
      name: 'apple',
      quantity: '50k',
      description: 'ladkald',
      price: 10000,
    };
    const products = await addNewProduct(newProduct);
    const throwError = () => {
      throw new TypeError(products);
    };
    expect(throwError).toThrow(`Error: quantity should be a type of 'number'`);
  });
});

describe('Product Service | Get Product', () => {
  it('should return all products available in array', async () => {
    const products = await getAllProduct();
    expect(products).toBeInstanceOf(Array);
  });

  it('should have valid objectID', async () => {
    const products = await getAllProduct();
    expect(products[0]).toHaveProperty('_id');
    expect(mongo.objectId.isValid(products[0])).toBeTruthy();
  });

  it('should return object of product', async () => {
    const products = await getAllProduct();
    const product = await getProductByProductId(products[0]._id);
    expect(product).toBeInstanceOf(Object);
  });

  it('should throw an error due to missing parameter productId', async () => {
    const product = await getProductByProductId();
    const throwError = () => {
      throw new TypeError(product);
    };
    expect(throwError).toThrow(`Error: id parameter is missing`);
  });

  it('should return null because productId not found', async () => {
    const product = await getProductByProductId(mongo.generateObjectId());
    expect(product).toBeNull();
  });
});
