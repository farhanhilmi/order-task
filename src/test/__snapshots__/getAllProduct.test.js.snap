// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Service | Add New Product should store a new product with correct property 1`] = `
Object {
  "__v": 0,
  "_id": "628dfb84a9d3adf920f5fbf3",
  "createdDate": 2022-05-25T09:48:52.840Z,
  "description": "ladkald",
  "modifyDate": 2022-05-25T09:48:52.840Z,
  "name": "apple",
  "price": 10000,
  "quantity": 50,
}
`;
